File: 141.环形链表.py
#
# @lc app=leetcode.cn id=141 lang=python3
#
# [141] 环形链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: 'ListNode') -> bool:
        cur = head
        visited = set()
        found = False
        while cur:
            if cur in visited:
                found = True
                break
            visited.add(cur)
            cur = cur.next
        return found


# @lc code=end

File: 141.环形链表.go
/*
 * @lc app=leetcode.cn id=141 lang=golang
 *
 * [141] 环形链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
	visited := make(map[*ListNode]int)
	found := false
	if head == nil {
		return false
	}
	for cur := head; cur != nil; cur = cur.Next {
		_, prs := visited[cur]
		if prs {
			found = true
			break
		}
		visited[cur] = 0
	}
	return found
}

// @lc code=end
